### ISSUE TITLE
a breaking change in 1.32 to scene.morphTo3D and scene.completeMorph

### ISSUE THREAD

**Issue by lftgly** [AnalyticalGraphicsInc/cesium#5197]
After updating from 1.31 to 1.32, code that toggles from 2D back to 3D using viewer.scene.morphTo3D(2.0) and scene.completeMorph() hangs-up the browser with either ballooning memory usage (and no error), or throws a rendering error, "DeveloperError: A PerspectiveFrustum or OrthographicFrustum is required in 3D and Columbus view."
After several frustrating hours narrowing-down the cause, it is related to
 Cesium 1.32 "Added support for an orthographic projection in 3D and Columbus view.
Set projectionPicker to true in the options when creating a Viewer to add a widget that will switch projections. #5021
Call switchToOrthographicFrustum or switchToPerspectiveFrustum on Camera to change projections."

A Sandcastle example to reproduce the error follows. Set the ScenePicker in the Viewer to 2D, then click the "3D" button. 
```
var viewer = new Cesium.Viewer('cesiumContainer');
Sandcastle.addToolbarButton('3D', function(){
switch (viewer.scene._mode){ 
case 2 :
 Cesium.when(viewer.scene.morphTo3D(2.0),function(){ 
 viewer.scene.completeMorph();
 console.log("scene.morphTo3D completed...");
 });
break; 
case 3 :
 console.log("already in 3D scene mode...");
break; 
}
});
```
When I paste this example code into http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Hello%20World.html&label=Showcases
change to 2D, and click "3D" button, it results in the following error:
An error occurred while rendering. Rendering has stopped.
An error occurred while rendering.  Rendering has stopped.
undefined
DeveloperError: A PerspectiveFrustum or OrthographicFrustum is required in 3D and Columbus view
Error
    at new DeveloperError (http://cesiumjs.org/Cesium/Source/Core/DeveloperError.js:44:19)
    at Camera.update (http://cesiumjs.org/Cesium/Source/Scene/Camera.js:922:19)
    at Scene.initializeFrame (http://cesiumjs.org/Cesium/Source/Scene/Scene.js:2519:22)
    at CesiumWidget.render (http://cesiumjs.org/Cesium/Source/Widgets/CesiumWidget/CesiumWidget.js:685:25)
    at render (http://cesiumjs.org/Cesium/Source/Widgets/CesiumWidget/CesiumWidget.js:72:32)

I do not get this error, or any error, on my development site!
The viewer begins to render the scene in 3D, then fails to load any more tiles, as TaskManager shows one of the chrome.exe processes eating up 2GB of available memory.

If  morphTo3D is allowed to complete, without forcing completeMorph, then morphTo3D does automatically switchToPerspectiveFrustum, from the 2D OrthographicOffCenterFrustum.

To minimalize this example for Sandcastle, I omitted the rest of the logic, which is a flyTo that should not  start until after the morphTo3D is complete.

Forcing completeMorph(); leaves the viewer in Cesium.OrthographicOffCenterFrustum and throws the error.

**Comment by hpinkos** [comment]
Thanks @lftgly!  @bagnell could you take a look at this?

**Comment by lftgly** [comment]
I will give you 3 more test cases. The 1st solves the problem by resetting a perspective frustum, which I did not have to do before Cesium v1.32. Run it, click the "3D" button to set up the view for fun, then use the ScenePicker to change to 2D, then click the "3D" button, flyTo works okay.

```
var viewer = new Cesium.Viewer('cesiumContainer');
Sandcastle.addToolbarButton('3D', function(){
    console.log("clicked 3D button...");
    var lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        } 
switch (viewer.scene._mode){ 
case 2 :
        console.log("Was in 2D scene mode...");
        console.log("scene.morphTo3D started...");
	Cesium.when(viewer.scene.morphTo3D(2.0),function(){	
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	
		lftglyFrustum = new Cesium.PerspectiveFrustum();
		lftglyFrustum.aspectRatio = viewer.scene.canvas.clientWidth / viewer.scene.canvas.clientHeight;
		lftglyFrustum.fov = Cesium.Math.PI_OVER_THREE;
		lftglyFrustum.near = 1.0;
		lftglyFrustum.far = 500000000.0;	//2.0;
		viewer.scene.camera.frustum = lftglyFrustum;	
		viewer.scene.completeMorph();
		console.log("scene.morphTo3D completed...");
		viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
        viewer.scene.camera.flyTo({
            destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
            maximumHeight : 14000,
            orientation : {
				heading : 5.0,
				pitch : -0.3,
				roll : 0.0}
        });
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	});
break; 
case 3 :
	console.log("already in 3D scene mode...");
        viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
		viewer.scene.camera.flyTo({
		destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
		maximumHeight : 14000,
		orientation : {
			heading : 5.0,
			pitch : -0.3,
			roll : 0.0}
		});	
break;	
}
});
```

2nd test case; Run it, click the "3D" button to set up the view for fun, then use the ScenePicker to change to 2D, then click the "3D" button, it throws the developer error. It's stuck in OrthographicOffCenterFrustum, if I do not programmatically reset the perspective frustum.

```
var viewer = new Cesium.Viewer('cesiumContainer');
Sandcastle.addToolbarButton('3D', function(){
    console.log("clicked 3D button...");
    var lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        } 
switch (viewer.scene._mode){ 
case 2 :
        console.log("Was in 2D scene mode...");
        console.log("scene.morphTo3D started...");
	Cesium.when(viewer.scene.morphTo3D(2.0),function(){	
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	
//		lftglyFrustum = new Cesium.PerspectiveFrustum();
//		lftglyFrustum.aspectRatio = viewer.scene.canvas.clientWidth / viewer.scene.canvas.clientHeight;
//		lftglyFrustum.fov = Cesium.Math.PI_OVER_THREE;
//		lftglyFrustum.near = 1.0;
//		lftglyFrustum.far = 500000000.0;	//2.0;
//		viewer.scene.camera.frustum = lftglyFrustum;	
		viewer.scene.completeMorph();
		console.log("scene.morphTo3D completed...");
		viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
        viewer.scene.camera.flyTo({
            destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
            maximumHeight : 14000,
            orientation : {
				heading : 5.0,
				pitch : -0.3,
				roll : 0.0}
        });
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	});
break; 
case 3 :
	console.log("already in 3D scene mode...");
        viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
		viewer.scene.camera.flyTo({
		destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
		maximumHeight : 14000,
		orientation : {
			heading : 5.0,
			pitch : -0.3,
			roll : 0.0}
		});	
break;	
}
});
```
A 3rd test case, exhibiting some strange behavior,  Run it, click the "3D" button to set up the view for fun, then use the ScenePicker to change to 2D, then click the "3D" button, Try clicking the "3D" button a second time.  If I reset the perspective frustum, but take out the completeMorph, then the fly To fails once, but works on second 3D button click. This may have more to do with the morphTo3D event I'm listening to, than the new frustum options.

```
var viewer = new Cesium.Viewer('cesiumContainer');
Sandcastle.addToolbarButton('3D', function(){
    console.log("clicked 3D button...");
    var lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        } 
switch (viewer.scene._mode){ 
case 2 :
        console.log("Was in 2D scene mode...");
        console.log("scene.morphTo3D started...");
	Cesium.when(viewer.scene.morphTo3D(2.0),function(){	
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	
		lftglyFrustum = new Cesium.PerspectiveFrustum();
		lftglyFrustum.aspectRatio = viewer.scene.canvas.clientWidth / viewer.scene.canvas.clientHeight;
		lftglyFrustum.fov = Cesium.Math.PI_OVER_THREE;
		lftglyFrustum.near = 1.0;
		lftglyFrustum.far = 500000000.0;	//2.0;
		viewer.scene.camera.frustum = lftglyFrustum;	
//		viewer.scene.completeMorph();
		console.log("scene.morphTo3D completed...");
		viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
        viewer.scene.camera.flyTo({
            destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
            maximumHeight : 14000,
            orientation : {
				heading : 5.0,
				pitch : -0.3,
				roll : 0.0}
        });
		lftglyFrustum = viewer.scene.camera.frustum;
        if (lftglyFrustum instanceof Cesium.PerspectiveFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveFrustum ");
        }
        if (lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum) {
            console.log("lftglyFrustum instanceof Cesium.PerspectiveOffCenterFrustum ");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicFrustum");
        }            
        if (lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum){
            console.log("lftglyFrustum instanceof Cesium.OrthographicOffCenterFrustum ");
        }	});
break; 
case 3 :
	console.log("already in 3D scene mode...");
        viewer.terrainProvider = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1]._creationCommand();
		viewer.baseLayerPicker.viewModel.selectedTerrain = viewer.baseLayerPicker.viewModel.terrainProviderViewModels[1];
		viewer.scene.camera.flyTo({
		destination : Cesium.Cartesian3.fromDegrees(-71.224917, 44.24931, 3000),
		maximumHeight : 14000,
		orientation : {
			heading : 5.0,
			pitch : -0.3,
			roll : 0.0}
		});	
break;	
}
});
```
Thanks for looking into this.
-Jon "Left Gully"

**Comment by bagnell** [comment]
This was fixed in #5262. Closing.

**Comment by mramato** [comment]
Thanks for the report @lftgly, this will be part of 1.33 being released in a couple of hours.
